### ISSUE TITLE
OSLC Architecture Management spec does not define any relationships that connect Resource to the rest of the lifecycle domains

### ISSUE THREAD

**Issue by jamsden** [oslc-op/oslc-specs#69]
The OSLC AM 2.1 spec rightfully defines oslc_am:Resource as a generic resource that could represent anything in the wide space of enterprise architecture, analysis and design. Servers are expected to subclass oslc_am:Resource to support more specific domains such as TOGAF, UML, SysML, Capella, etc. 

However, OSLC AM 2.1 does not define any relationships between oslc_am:Resource and any of the other OSLC lifecycle classes. I talked with Jim Conallen about this and he indicated that the primary reason for this was the variability in enterprise architecture, analysis and design methods, notations and tools made it difficult to standardize on any particular set of properties. 

So instead OSLC AM 2.0 defined oslc_am:LinkType to provide a discoverable way for particular OSLC servers to describe their supported links, and also include properties on links. However, LinkType was never implemented in any IBM tool and the open-services.net AM workgroup proposed removing LinkType from the OSLC AM 3.0 specification. A possible reason for this is that LinkType, like reified RDF statements do not guarantee assertion of the described link, and standard OWL reasoners don't infer these assertions either (and of course OSLC specifically avoids reasoners). Also queries between AM resources and other elements would be complex and inefficient. And the infrastructure supporting links would be different for AM resources than other OSLC domain classes.

Instead what happened is a few properties were added to the RM 2.0 spec (e.g., oslc_rm:satisfies) that could link an AM resource to a requirement, and IBM Design Manager defined set of properties it supported in https://jazz.net/wiki/bin/view/LinkedData/DmLinkTypesVocabulary (e.g., Derives From, Elaborates, External Web Page, Refines, Satisfies, Trace). However, for whatever reason, no similar properties that connect AM resources to other lifecycle domain classes were ever defined in the AM 2.0 spec. And the AM 2.1 spec makes no mention at all of connections to requirements, change requests, test cases, or other resources. 

This seems wrong. AM resources don't seem like a lifecycle afterthought, they seem like the point of lifecycle management. That is an AM resource would seem to represent the thing whose lifecycle is being managed! Having requirements, change requests and test cases that connect to nothing that is actually delivered seems rather pointless. Not being able to connect such a fundamental concept to the rest of the OSLC domain classes seems like a glaring omission, that perhaps happened by accident of history. 

The [Architecture Management 3.0](https://archive.open-services.net/wiki/architecture-management/OSLC-Architecture-Management-Specification-Version-3.0/index.html#Resource_Resource) draft specification did include properties to connect AM resources to other lifecycle classes including:

* oslc_am:refines
* oslc_am:satisfies
* oslc_am:verifies
* dm:derives
* dm:elaborates

These properties were introduced to support SysML and the RDM implementation. Presumably the dm namespace would be replaced by oslc_am, but wasn't in this draft spec.

What I'd like to suggest is that we consider correcting this situation by defining a small set of generic properties that link AM resources to other lifecycle resources. I propose the following:

* oslc_am:Resource dcterms:relation oslc_am:Resource
* oslc_am:Resource oslc_rm:satisfies oslc_rm:Requirement
* oslc_am:Resource oslc_rm:implementedBy oslc_cm:ChangeRequest
* oslc_am:Resource oslc_rm:validatedBy oslc_qm:TestCase

I know it's a little strange to reuse oslc_rm properties in the AM spec. But 1) they're more or less the right generic names 2) we use properties defined in other namespaces all the time (e.g., dcterms:title) and 3) we can provide links between AM resources and the rest of the lifecycle without introducing any new properties.

**Comment by berezovskyi** [comment]
> A possible reason for this is that LinkType, like reified RDF statements do not guarantee assertion of the described link, and standard OWL reasoners don't infer these assertions either (and of course OSLC specifically avoids reasoners)

This is pure contradiction. I think we agree that OWL support is not required (as opposed to be avoided) for OSLC applications. Therefore, custom OSLC semantics apply to LinkType, which are to be enforced programmatically as anything else in OSLC.

**Comment by berezovskyi** [comment]
> Also queries between AM resources and other elements would be complex and inefficient. And the infrastructure supporting links would be different for AM resources than other OSLC domain classes.

I have recently learned about `rdf:value`. The idea is that for any property p, we should assume `p/rdf:value` is the same as p. This should allow to implement the links efficiently, putting a LinkType into the blank node with the `rdf:value`. Given that you use a triplestore and not an SQL database inside your app. SPARQL 1.1 makes writing such queries very simple: https://www.w3.org/TR/sparql11-query/#pp-language (e.g. `?s oslc_am:refines/rdf:value ?o` to get the link)

**Comment by berezovskyi** [comment]
> However, for whatever reason, no similar properties that connect AM resources to other lifecycle domain classes were ever defined in the AM 2.0 

seems to be explained by 

> Jim Conallen about this and he indicated that the primary reason for this was the variability in enterprise architecture

**Comment by berezovskyi** [comment]
> I know it's a little strange to reuse oslc_rm properties in the AM spec. 

I like the idea of reusing the properties. We just need to be careful to ensure nobody assumed that the domain of `oslc_rm:implementedBy` and `oslc_rm:validatedBy` must be a Requirement.

**Comment by berezovskyi** [comment]
> we can provide links between AM resources and the rest of the lifecycle without introducing any new properties.

We need to think about not introducing ambiguities instead of not introducing new properties.

**Comment by berezovskyi** [comment]
> IBM Design Manager defined set of properties it supported in https://jazz.net/wiki/bin/view/LinkedData/DmLinkTypesVocabulary (e.g., Derives From, Elaborates, External Web Page, Refines, Satisfies, Trace)

I think we should incorporate them in our specs. We can also define some of the properties as superproperties of those, e.g.

```
am:derives a rdfs:Property.

http://jazz.net/ns/dm/linktypes#derives rdfs:subPropertyOf am:derives .
```

This would  not be natural but semantically correct and hey, AAA!

**Comment by jamsden** [comment]
Interestingly AM 2.0 LinkType never defined the source and target property types. So this was pretty limited information anyway.

The draft AM 3.0 spec on open-services.net has some suggested properties. We should look at those too, as well as what was actually implemented in Rational Design Manager, and what is planned to be implemented in Rational Model Manager.

**Comment by jamsden** [comment]
We agreed that the current AM 2.1 spec is a good starting point, but it is sufficiently incomplete in its current for to not be worth progressing to an OASIS Standard. We should proceed with AM 3.0 in order to address this issue.

We need to look at the properties that were added to the draft AM 3.0 spec on open-services.net, and what has been implemented in practice, and develop an AM 3.0 spec that properly positions AM resources into the broader lifecycle management domain.

I have reviewed the old OSLC AM 3.0 draft, the Rational Design Manager vocabulary published on jazz.net/ns, and requested information about Rational Model Manager linking to CM, RM and QM resources. Once we gather all this information, we can hopefully be in a better possition to know how to proceed with the AM spec.

**Comment by jamsden** [comment]
Some additional notes regarding Rational Model Manager's (RMM) implementation:

RMM preserves the jazz.net DM namespace http://jazz.net/ns/dm/linktypes# and the derives and elaborates properties defined in that namespace. So did the draft AM 3.0 spec.

RMM supports linking to QM TestCase using http://jazz.net/ns/qm/rqm#validatesArchitectureElement - that is, the link is defined in a jazz.net extension to QM.  See https://github.com/oslc-op/oslc-specs/wiki/Automation-Specification-Notes and https://github.com/oslc-op/oslc-specs/wiki/Actions-Specification-Strategy. 

Basically we have limited dependencies on Actions and pretty much put it on hold. So it should be retained, and not referenced, at least not yet.
